nodist/clang_check_attributes

```diff
@@ -0,0 +1,332 @@
+#!/usr/bin/env python3
+# pylint: disable=C0114,C0115,C0116,C0209
+#
+# Copyright 2022 by Wilson Snyder. Verilator is free software; you
+# can redistribute it and/or modify it under the terms of either the GNU Lesser
+# General Public License Version 3 or the Apache License 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Apache-2.0
+#
+# Based on the work of (Guillaume "Vermeille" Sanchez)[https://github.com/Vermeille/clang-callgraph]
+# licensed under Apache 2.0
+
+import argparse
+import sys
+import os
+from pathlib import Path
+from pprint import pprint
+from clang.cindex import CursorKind, Index
+
+
+# https://stackoverflow.com/a/74205044
+class FixSizedDict(dict):
+
+    def __init__(self, *args, maxlen=0, **kwargs):
+        self._maxlen = maxlen
+        super().__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        dict.__setitem__(self, key, value)
+        if self._maxlen > 0:
+            if len(self) > self._maxlen:
+                self.pop(next(iter(self)))
+
+
+# by definition constructor is MT safe,
+# but functions that constructor call are not!
+MT_SAFE_KINDS = [CursorKind.CONSTRUCTOR]
+VERILATOR_ROOT = Path(__file__).parent.parent.resolve()
+IGNORE_FILES_LIST = [
+    # Dont process test files
+    "_test.cpp",
+    # Dont process Verilator coverage files
+    "Vlc",
+    # We want to use V3Const after astgen
+    "V3Const.cpp",
+    # Skip SystemC files
+    "_sc",
+    # Skip parser files
+    ".yy.cpp"
+]
+
+index = Index.create()
+# Cache of files already turned into translate unit
+# this speeds up checking when used function is from another
+# translation unit, but it increases memory consumption.
+# Limit max cache size to 20, to make sure we won't run OOM
+# in github actions
+translate_units = FixSizedDict(maxlen=20)
+# Already printed nodes in current function.
+# This prevents infinite loops due to recursion
+printed = []
+# Nodes that were marked as MT_SAFE, but are calling
+# function that isn't marked with this attribute
+unsafe_calls = []
+
+
+def get_diag_info(diag):
+    return {
+        'severity': diag.severity,
+        'location': diag.location,
+        'spelling': diag.spelling,
+        'ranges': list(diag.ranges),
+        'fixits': list(diag.fixits)
+    }
+
```

Comment:
```suggestion
```

Please put single line between functions, only double line is where "make format" forces them.

Reply:
``make format`` forces double line between functions.